flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java

```diff
@@ -137,7 +138,12 @@ public JobManagerRunnerImpl(
 		this.leaderGatewayFuture = new CompletableFuture<>();
 
 		// now start the JobManager
-		this.jobMasterService = jobMasterFactory.createJobMasterService(jobGraph, this, userCodeLoader);
+		try {
+			this.jobMasterService = jobMasterFactory.createJobMasterService(jobGraph, this, userCodeLoader);
+		} catch (Throwable t) {
+			ClusterEntryPointExceptionUtils.tryEnrichClusterEntryPointError(t);
+			throw t;
+		}
```

Comment:
I would move this specific exception handling out because now we would only enrich OOMs which occur here. But what if an OOM happens in a line above (e.g. `classLoaderLease.getOrResolveClassLoader(...)`)? I think it would be better to put it into `Dispatcher.internalSubmitJob`.

Reply:
I see. But wouldn't we miss doing the error handling for a Job rerun. The JobManager instantiation happens in `Dispatcher.persistAndRun(JobGraph)` (which calls `Dispatcher.internalSubmitJob(JobGraph)` but also in `Dispatcher.runRecoveredJob(JobGraph)`. The latter call graph does not touch `Dispatcher.internalSubmitJob(JobGraph)` which means that we wouldn't be able to handle the OOM for Job reruns if moving it into `Dispatcher.internalSubmitJob(JobGraph)` in that case.

Wouldn't `Dispatcher.runJob(JobGraph)` the better location for the OOM handling?

Reply:
`Dispatcher.runRecoveredJob` will call `handleRecoveredJobStartError` if there is an error. This method will call the `ClusterEntrypoint's` fatal error handler which should already do the enrichment.

Reply:
👍  I moved the code into `Dispatcher.internalSubmitJob(Graph)`.